### PR TITLE
Fix the performance issue

### PR DESCRIPTION
**Pull request by szc321**
Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
-   [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
-   [ ] Tests are part of the PR (for bug fixes / features)
-   [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
-   [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
-   [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
-   [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
-   [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

-   [ ] Bugfix
-   [ ] Feature
-   [ ] Code style update (formatting, renaming)
-   [ ] Refactoring (no functional changes, no api changes)
-   [ ] Build related changes
-   [ ] Documentation changes
-   [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

-   [ ] Yes
-   [ ] No

### CODE REVIEW

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/message_infrastructure_py_wrapper.cc [hunk 1]
```diff
@@ -215,6 +215,9 @@ PYBIND11_MODULE(MessageInfrastructurePywrapper, m) {
     .def_property_readonly("shape", &RecvPortProxy::Shape)
     .def_property_readonly("d_type", &RecvPortProxy::DType)
     .def_property_readonly("size", &RecvPortProxy::Size);
+  py::class_<Selector, std::shared_ptr<Selector>> (m, "CPPSelector")
+  .def(py::init<>())
```
Comment (hexu33) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/message_infrastructure_py_wrapper.cc:219]: Modify code indentation

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc [hunk 2]
```diff
@@ -224,4 +233,39 @@ py::object RecvPortProxy::MDataToObject_(MetaDataPtr metadata) {
...
+
+
+void Selector::Changed() {
+    std::unique_lock<std::mutex> lock(cv_mutex_);
+    ready_ = true;
+    cv_.notify_all();
+}
+
+void Selector::Set_observer(std::vector<std::tuple<RecvPortProxyPtr,
+                      py::function>> *channel_actions,
```
Comment (hexu33) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc:245]: Modify code indentation

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc [hunk 3]
```diff
@@ -224,4 +233,39 @@ py::object RecvPortProxy::MDataToObject_(MetaDataPtr metadata) {
...
+          it != channel_actions->end(); ++it) {
+        std::get<0>(*it)->Set_observer(observer);
+    }
+}
+
+pybind11::object Selector::Select(std::vector<std::tuple<RecvPortProxyPtr,
+                              py::function>> *args) {
+  std::function<void()> observer = std::bind(&Selector::Changed, this);
+  Set_observer(args, observer);
+    while (true) {
```
Comment (hexu33) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc:257]: Modify code indentation

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc [hunk 4]
```diff
@@ -224,4 +233,39 @@ py::object RecvPortProxy::MDataToObject_(MetaDataPtr metadata) {
   return py::reinterpret_steal<py::object>(array);
 }
 
+
+
+void Selector::Changed() {
+    std::unique_lock<std::mutex> lock(cv_mutex_);
```
Comment (hexu33) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc:239]: Modify code indentation

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.h [hunk 5]
```diff
@@ -73,6 +77,7 @@ class RecvPortProxy : public PortProxy {
   py::object Peek();
   std::string Name();
   size_t Size();
+  void Set_observer(std::function<void()> obs);
```
Comment (hexu33) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.h:80]: Follow func name rule.

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.h [hunk 6]
```diff
@@ -87,6 +92,24 @@ using RecvPortProxyPtr = std::shared_ptr<RecvPortProxy>;
...
+class Selector {
+ private:
+  std::condition_variable cv_;
+  mutable std::mutex cv_mutex_;
+  bool ready_ = false;
+
+ public:
+  void Changed();
+
+  void Set_observer(std::vector<std::tuple<RecvPortProxyPtr,
```
Comment (hexu33) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.h:105]: Follow func name rule.

File: src/lava/magma/runtime/message_infrastructure/ports.py [hunk 7]
```diff
@@ -18,17 +18,16 @@
 import numpy as np
 import typing as ty
 import warnings
+from lava.magma.runtime.message_infrastructure.MessageInfrastructurePywrapper \
+    import CPPSelector
+
 
+class Selector(CPPSelector):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/ports.py:25]: move it into `__init__.py` as following:
```py
from lava.magma.runtime.message_infrastructure.MessageInfrastructurePywrapper \
    import CPPSelector as Selector
```

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc [hunk 8]
```diff
@@ -167,6 +167,15 @@ size_t RecvPortProxy::Size() {
   return recv_port_->Size();
 }
 
+void RecvPortProxy::SetObserver(std::function<void()> obs) {
+  recv_port_->obs_lk_.lock();
```
Comment (killight98) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.cc:171]: replace by [`std::lock_guard`](https://en.cppreference.com/w/cpp/thread/lock_guard)

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/channel/shmem/shmem_port.cc [hunk 9]
```diff
@@ -99,15 +99,15 @@ void ShmemRecvPort::QueueRecv() {
   while (!done_.load()) {
     bool ret = false;
     if (this->recv_queue_->AvailableCount() > 0) {
-      ret = shm_->Load([this](void* data){
+      bool not_empty = recv_queue_->Probe();
```
Comment (killight98) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/channel/shmem/shmem_port.cc:102]: useless?

File: src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.h [hunk 10]
```diff
@@ -73,6 +77,7 @@ class RecvPortProxy : public PortProxy {
   py::object Peek();
   std::string Name();
   size_t Size();
+  void SetObserver(std::function<void()> obs);
```
Comment (killight98) [inline src/lava/magma/runtime/_c_message_infrastructure/csrc/port_proxy.h:80]: useless?

**Review by killight98 — CHANGES_REQUESTED** [review]

**Review by killight98 — APPROVED** [review]

**Pull request merged**